### ISSUE TITLE
boilerplate for sample lodash type utility library

### ISSUE THREAD

**Issue by nsisodiya** [developit/microbundle#597]
is microbundle has any boilerplate which can create something like this?

Requirement
======
I am looking for near-perfect boilerplate for creating lodash type library. boilerplate should have the following features.

* source code is written in typescript but output files/bundles should be able to be used in the project which use typescript or javascript.
* It should generate sufficient bundles (and files) so that it can be used in 

*  [Without bundler] [Browser] Browser via script tag with hosted cdnjs URL (min+uglify). expose only one global variable Ex - myTsMathlib.calCircleArea(4), myTsMathlib.calSquareArea()
*  [Without bundler] [Browser] can be included as script tag and type module.
*  [Without bundler] [Node] Node js via require statements like  
```
var { calCircleArea } = require('my-ts-mathlib');
```
or

```
const calCircleArea = require('my-ts-mathlib/es/calCircleArea');
```
* [With Bundler] [with or without typescript] [Webpack or CRA]  If library has 100+ function, only those function which are included in the source code should go in the bundle. we should be able to include like this
```
import calCircleArea from 'my-ts-mathlib/es/calCircleArea';
```

**Comment by nsisodiya** [comment]
I have created a utility library and published over npm.
https://github.com/nsisodiya/my-ts-mathlib
I have included various examples in the example folder.

**Comment by marvinhagemeister** [comment]
There is no need for lodash-style duplication of packages. The standard way of doing it is setting up both `main` and `module` fields correctly in `package.json` and the rest will be taken care of.

With that users will be able to consume your package in the standard way, instead of having to workarounds for each botched package. Seeing so many people get it wrong was one of the reasons extracted out build tooling from Preact and published it as `microbundle`.

```js
// package.json
{
  "main": "dist/foo.js",
  "module": "dist/foo.module.js"
}
```

And with just that your users will be able to use your library like they expect it to be used:

```js
// CommonJS
var { callCircleArea } = require('my-ts-mathlib');

// ES Modules
import { callCircleArea } from 'my-ts-mathlib';
```

**Comment by nsisodiya** [comment]
@marvinhagemeister Thanks a lot for your reply.
I have few questions.
1. name convention.
```
"module": "dist/foo.module.js"
or
"module": "dist/foo.mjs"
```
`node experimental modules` use mjs extension for files. So, I am confused about this part. what is the right naming convention?

2. TreeShaking or No unwanted Code in the bundle.

above utility, the library has calCircleArea and calSquareArea. 
When I used calSquareArea in CRA code, somehow magically, the final bundle doesn't include the code for 'calCircleArea' function which was not used.
But when I passed the same code to parcel-bundler, it included all functions.

I just want to make sure, the user should be able to include individual functions.

3. import statement.
Browser now supports type="module" and for using it, I have to do it like this.
```
import { calSquareArea } from './node_modules/my-ts-mathlib/dist/index.module.js';
```
This is the case of no bundler. https://github.com/nsisodiya/my-ts-mathlib/blob/master/example/brower-es-modules/main.js
This code is working but the only problem, import statement is a bit lengthy.

**Comment by marvinhagemeister** [comment]
@nsisodiya Sorry for the late reply. Missed the notification for this in GitHub. To answer your questions:

1) name convention

Doesn't matter for the module field. Using `.mjs` as the extension breaks bundling with webpack though. They don't resolve any `.mjs` files last time I checked.

2) TreeShaking or No unwanted Code in the bundle.

Tree-Shaking is a variant of Dead-Code-Elimination (short: DCE). Basically it means the function that got removed were inaccessible to begin with. The entry bundle defines what is marked as used and what not.

```js
// Used by bob().
function foo() {}

// Not used anywhere, will be DCE'd
function bar() {}

// Top level exports are in the entry file are always marked as used
export function bob() {
  return foo();
}
```

3) import statement. Browser now supports type="module" and for using it, I have to [use absolute urls].

Yep, browsers resolve modules by URL, not by the node resolution algorithm that is common with bundlers and in node. There is no alternative right now, it's in my opinion a gap in the spec. There was some talk about import maps, but nothing got traction so far.
